File: src-tauri/src/commands/features.rs
use crate::domain::ids::{FeatureId, StepExecutionId};
use crate::domain::models::{
    EffortLevel, Feature, GateDecision, SequenceState, StepAttempt, StepExecution,
};
use crate::error::AppError;
use crate::ports::step_executor::{FeatureLaunch, SyncOutcomeView};
use crate::ports::sync_session::SyncSession;
use crate::state::AppContext;
use tauri::State;

// Re-exported so the wire shape the frontend already consumes is unchanged;
// the resolution logic now lives in `demeteo_core` so the runner's
// `get_worktree` control RPC shares exactly one code path with this command.
pub use crate::application::worktree::FeatureWorktreeInfo;

#[tauri::command]
pub async fn feature_get_worktree(
    ctx: State<'_, AppContext>,
    feature_id: String,
) -> Result<FeatureWorktreeInfo, AppError> {
    crate::application::worktree::resolve_feature_worktree(&ctx, &FeatureId::from(feature_id))
        .await
        .map_err(AppError::from)
}

#[tauri::command]
pub fn fetch_active_features(
    ctx: State<'_, AppContext>,
    project_id: String,
) -> Result<Vec<Feature>, AppError> {
    ctx.features
        .get_active(&crate::domain::ids::ProjectId::from(project_id))
        .map_err(AppError::from)
}

/// Pre-launch user attachments. Persisted to the feature row BEFORE
/// the driver is spawned so the agent's first turn sees them. Pass
/// `None` (or an empty array) when the user did not attach anything.
#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub async fn start_feature(
    ctx: State<'_, AppContext>,
    project_id: String,
    workflow_id: String,
    title: String,
    description: String,
    agent_kind: Option<String>,
    model: Option<String>,
    // Feature-wide effort. Omitted (an older frontend) = `None` = inherit the
    // project default, which bottoms out at `EffortLevel::DEFAULT`. Per-step
    // efforts ride inside `step_overrides`, not here.
    effort: Option<EffortLevel>,
    commit_artifacts: Option<bool>,
    loop_iterations: Option<u32>,
    // Per-run dollar budget override (`--max-budget-usd`). Omitted (an older
    // frontend) = `None` = inherit the project default, then the engine
    // default.
    max_budget_usd: Option<f64>,
    step_overrides: Option<Vec<crate::domain::models::StepOverride>>,
    staged_attachments: Option<Vec<crate::commands::attachments::StagedAttachmentInput>>,
    // Omitted (a frontend older than the origin picker) = `None`, which
    // `FeatureLaunch::origin` and `FeatureLaunch::diff_base_branch` define.
    origin: Option<crate::domain::feature_origin::FeatureOrigin>,
    diff_base_branch: Option<String>,
) -> Result<Feature, AppError> {
    ctx.executor
        .feature_start(FeatureLaunch {
            project_id,
            workflow_id,
            title,
            description,
            agent_kind,
            model,
            effort,
            commit_artifacts,
            loop_iterations,
            max_budget_usd,
            step_overrides: step_overrides.unwrap_or_default(),
            staged_attachments: staged_attachments.unwrap_or_default(),
            origin: origin.unwrap_or_default(),
            diff_base_branch,
            ..FeatureLaunch::default()
        })
        .await
        .map_err(AppError::from)
}

#[tauri::command]
pub async fn feature_pause(ctx: State<'_, AppContext>, feature_id: String) -> Result<(), AppError> {
    ctx.executor
        .feature_pause(&feature_id)
        .await
        .map_err(AppError::from)
}

#[tauri::command]
pub async fn feature_resume(
    ctx: State<'_, AppContext>,
    feature_id: String,
) -> Result<(), AppError> {
    ctx.executor
        .feature_resume(&feature_id)
        .await
        .map_err(AppError::from)
}

#[tauri::command]
pub async fn feature_cancel(
    ctx: State<'_, AppContext>,
    feature_id: String,
) -> Result<(), AppError> {
    ctx.executor
        .feature_cancel(&feature_id)
        .await
        .map_err(AppError::from)
}

#[tauri::command]
pub async fn step_get(
    ctx: State<'_, AppContext>,
    execution_id: String,
) -> Result<StepExecution, AppError> {
    // Read-model path (C3): display reads go through `RunView`, not the
    // executor, so a runner-owned step can later resolve from the shadow.
    ctx.run_view
        .step(&StepExecutionId::from(execution_id))
        .map_err(AppError::from)?
        .ok_or_else(|| AppError::not_found("Step execution not found".to_string()))
}

/// Per-attempt history for a step execution — the node drill-down panel's
/// Overview tab (P2.3). Read-model path (C3) like `step_get`: goes through
/// `RunView`, so a runner-owned step's attempts resolve from the shadow.
#[tauri::command]
pub async fn step_attempts_list(
    ctx: State<'_, AppContext>,
    execution_id: String,
) -> Result<Vec<StepAttempt>, AppError> {
    ctx.run_view
        .step_attempts(&StepExecutionId::from(execution_id))
        .map_err(AppError::from)
}

/// A `sequence` node's task list for the drill-down accordion (P2.5): each
/// task's landed-vs-pending state (Decision 13's committed prefix) plus its
/// per-task cost. Read-model path (C3) like `step_attempts_list`. `node_id` is
/// the graph node id (== v1 `step_id`); `execution_id` is its step-execution
/// row. A non-sequence or not-yet-planned node reads back `unplanned`.
#[tauri::command]
pub async fn sequence_tasks_list(
    ctx: State<'_, AppContext>,
    feature_id: String,
    node_id: String,
    execution_id: String,
) -> Result<SequenceState, AppError> {
    ctx.run_view
        .sequence_state(
            &FeatureId::from(feature_id),
            &node_id,
            &StepExecutionId::from(execution_id),
        )
        .map_err(AppError::from)
}

#[tauri::command]
pub async fn step_list_for_run(
    ctx: State<'_, AppContext>,
    feature_id: String,
) -> Result<Vec<StepExecution>, AppError> {
    ctx.run_view
        .steps(&FeatureId::from(feature_id))
        .map_err(AppError::from)
}

#[tauri::command]
pub async fn gate_pending_for_run(
    ctx: State<'_, AppContext>,
    feature_id: String,
) -> Result<Option<GateDecision>, AppError> {
    ctx.presenter
        .gate_pending_for_run(&FeatureId::from(feature_id))
        .await
        .map_err(AppError::from)
}

#[tauri::command]
pub async fn gate_decide(
    ctx: State<'_, AppContext>,
    step_execution_id: String,
    decision: String,
    feedback: Option<String>,
) -> Result<(), AppError> {
    ctx.presenter
        .gate_decide(
            &StepExecutionId::from(step_execution_id),
            &decision,
            feedback.as_deref(),
        )
        .await
}

#[tauri::command]
pub async fn step_retry(
    ctx: State<'_, AppContext>,
    step_execution_id: String,
    new_model: Option<String>,
    new_agent: Option<String>,
    // Re-pin the feature-wide effort for the rerun, exactly as `new_model`
    // re-pins the model. `None` (or an older frontend that omits it) keeps
    // whatever the feature already carries.
    new_effort: Option<EffortLevel>,
) -> Result<(), AppError> {
    ctx.executor
        .step_retry(
            &step_execution_id,
            new_model.as_deref(),
            new_agent.as_deref(),
            new_effort,
        )
        .await
}

#[tauri::command]
pub async fn replay_from_step(
    ctx: State<'_, AppContext>,
    step_execution_id: String,
    new_model: Option<String>,
    new_agent: Option<String>,
    new_effort: Option<EffortLevel>,
) -> Result<(), AppError> {
    ctx.executor
        .replay_from_step(
            &step_execution_id,
            new_model.as_deref(),
            new_agent.as_deref(),
            new_effort,
        )
        .await
        .map_err(AppError::from)
}

#[tauri::command]
pub fn feature_get(
    ctx: State<'_, AppContext>,
    feature_id: String,
) -> Result<Option<Feature>, AppError> {
    ctx.run_view
        .feature(&FeatureId::from(feature_id))
        .map_err(AppError::from)
}

/// The UTF-8 body of a run's declared artifact. This is a *display* read of a
/// run surface, so it goes through `RunView` (C3) rather than `sftp_read_file`
/// directly — that is the seam C4 uses to serve a runner-owned feature's
/// artifact from the lazily-cached laptop shadow. General filesystem browsing
/// (the code editor) stays on `sftp_read_file`; only run-artifact display uses
/// this.
#[tauri::command]
pub async fn artifact_body(
    ctx: State<'_, AppContext>,
    machine_id: String,
    path: String,
) -> Result<String, AppError> {
    ctx.run_view
        .artifact_body(&machine_id, &path)
        .await
        .map_err(AppError::from)
}

/// Sync the feature branch with `origin/<default_branch>`. Returns
/// a `SyncOutcomeView` the UI can render directly:
/// - `Ok` when the merge was clean (or there was nothing to merge).
/// - `Conflict` when the merge left unmerged files; the UI offers a
///   "Resolve with agent" button that calls
///   `feature_resolve_sync_conflicts` with the same conflict list.
/// - `Blocked` when the sync stopped short of a merge; there is
///   nothing for an agent to resolve.
/// - `Resolved` after a successful agent resolution.
#[tauri::command]
pub async fn feature_sync(
    ctx: State<'_, AppContext>,
    feature_id: String,
) -> Result<SyncOutcomeView, AppError> {
    ctx.executor
        .feature_sync(&feature_id)
        .await
        .map_err(AppError::from)
}

/// The feature's live sync, or `null` if it has never synced.
///
/// Reconciled against the working tree before it answers, so a session left
/// `resolving` by a process that died — or a conflict whose worktree a later
/// sync force-removed — is never handed to the UI as if it were still true.
#[tauri::command]
pub async fn sync_session_get(
    ctx: State<'_, AppContext>,
    feature_id: String,
) -> Result<Option<SyncSession>, AppError> {
    crate::application::sync_session::get_reconciled(
        &ctx.sync_sessions,
        &ctx.exec,
        &FeatureId::from(feature_id),
    )
    .await
    .map_err(AppError::from)
}

/// Give up on the feature's sync: undo the merge, discard the worktree, and
/// close the session. Safe on a worktree that is already gone, which is the
/// common case.
#[tauri::command]
pub async fn sync_abort(
    ctx: State<'_, AppContext>,
    feature_id: String,
) -> Result<Option<SyncSession>, AppError> {
    crate::application::sync_session::abort(
        &ctx.sync_sessions,
        &ctx.exec,
        &FeatureId::from(feature_id),
    )
    .await
    .map_err(AppError::from)
}

/// Spawn a fresh agent to resolve the conflicts left by
/// `feature_sync`. The agent edits the conflict files in a temporary
/// worktree, commits the resolution, and the worktree is merged back
/// into the feature branch.
#[tauri::command]
pub async fn feature_resolve_sync_conflicts(
    ctx: State<'_, AppContext>,
    feature_id: String,
    conflict_files: Option<Vec<String>>,
) -> Result<SyncOutcomeView, AppError> {
    let files = conflict_files.unwrap_or_default();
    ctx.executor
        .feature_resolve_sync_conflicts(&feature_id, &files)
        .await
        .map_err(AppError::from)
}

File: src/lib/features.ts
import { invoke } from "@tauri-apps/api/core";
import { asAppError } from "./errors";
import type {
  EffortLevel,
  Feature,
  SequenceState,
  StepAttempt,
  StepExecution,
} from "../types";

/** Features still executing (or awaiting a gate) in a project. */
export async function fetchActiveFeatures(projectId: string): Promise<Feature[]> {
  return invoke<Feature[]>("fetch_active_features", { projectId });
}

/** Where a feature's worktree lives, resolved at call time. Mirrors the Rust
 *  `FeatureWorktreeInfo`, which the runner's `get_worktree` RPC returns too —
 *  so a caller opening a terminal on it needs no transport branch. */
export interface FeatureWorktreeInfo {
  machine_id: string;
  worktree_path: string;
  branch: string;
  default_branch: string;
}

export async function getFeatureWorktree(featureId: string): Promise<FeatureWorktreeInfo> {
  return invoke<FeatureWorktreeInfo>("feature_get_worktree", { featureId });
}

export async function getStepExecution(executionId: string): Promise<StepExecution> {
  return invoke<StepExecution>("step_get", { executionId });
}

export async function listStepsForRun(featureId: string): Promise<StepExecution[]> {
  return invoke<StepExecution[]>("step_list_for_run", { featureId });
}

/** Per-attempt history for one step execution. */
export async function listStepAttempts(executionId: string): Promise<StepAttempt[]> {
  return invoke<StepAttempt[]>("step_attempts_list", { executionId });
}

/** A `sequence` node's task list, split into the landed prefix and what is
 *  still pending. `nodeId` is the graph node id; a non-sequence or
 *  not-yet-planned node reads back `unplanned`. */
export async function getSequenceState(input: {
  featureId: string;
  nodeId: string;
  executionId: string;
}): Promise<SequenceState> {
  return invoke<SequenceState>("sequence_tasks_list", {
    featureId: input.featureId,
    nodeId: input.nodeId,
    executionId: input.executionId,
  });
}

/**
 * The UTF-8 body of a run's declared artifact. A *display* read of a run
 * surface, so it goes through `RunView` rather than `sftp_read_file` — the
 * seam that lets a runner-owned feature's artifact resolve from the laptop
 * shadow. General filesystem browsing stays on `lib/files.ts`.
 */
export async function artifactBody(machineId: string, path: string): Promise<string> {
  return invoke<string>("artifact_body", { machineId, path });
}

/**
 * Subset of `StepExecution` the gate-modal block-banner actually needs.
 * Kept narrow so the type can be reused by callers that only fetch the
 * predecessor set (without dragging the whole row in).
 */
export type GateBlocker = Pick<StepExecution, "id" | "step_id" | "status" | "step_index">;

/**
 * Phrases the backend uses to describe a blocked retry / gate decision.
 * Kept in lock-step with the `assert_no_active_predecessors` helper in
 * `src-tauri/src/adapters/step_executor/impl_traits/mod.rs`. Used by
 * `isBlockingError` to detect the precondition violation regardless of
 * whether the error surface carries `AppError::validation` (preferred)
 * or the legacy string-only path.
 */
export const BLOCKING_ERROR_PHRASES = [
  "is still pending",
  "is still running",
  "is still verifying",
  "is still awaiting",
] as const;

/**
 * Decide a manual gate step (`approve | redirect | cancel`). Wraps
 * the raw `invoke('gate_decide', …)` call so the UI has one
 * centralised place to detect the blocking-predecessor error and
 * surface it with a tailored toast.
 */
export async function decideGate(input: {
  stepExecutionId: string;
  decision: "approve" | "redirect" | "cancel";
  feedback: string | null;
}): Promise<void> {
  await invoke<void>("gate_decide", {
    stepExecutionId: input.stepExecutionId,
    decision: input.decision,
    feedback: input.feedback,
  });
}

/**
 * Retry a failed / interrupted / pending step. Re-pins the
 * feature-wide model / harness / effort overrides before the rerun
 * (`null` keeps the existing override).
 */
export async function retryStep(input: {
  stepExecutionId: string;
  newModel: string | null;
  newAgent: string | null;
  newEffort: EffortLevel | null;
}): Promise<void> {
  await invoke<void>("step_retry", {
    stepExecutionId: input.stepExecutionId,
    newModel: input.newModel,
    newAgent: input.newAgent,
    newEffort: input.newEffort,
  });
}

/**
 * Rewind to a step and re-execute it plus everything downstream, re-pinning
 * the same three overrides as {@link retryStep}.
 *
 * Close to {@link retryStep} but not the same call, and the differences bite:
 * this one accepts a step of *any* status (a replay target is normally
 * `completed`) and drops a sequence step's landed checkpoint so it runs its
 * whole task list again.
 */
export async function replayFromStep(input: {
  stepExecutionId: string;
  newModel: string | null;
  newAgent: string | null;
  newEffort: EffortLevel | null;
}): Promise<void> {
  await invoke<void>("replay_from_step", {
    stepExecutionId: input.stepExecutionId,
    newModel: input.newModel,
    newAgent: input.newAgent,
    newEffort: input.newEffort,
  });
}

/**
 * The detached twin of {@link retryStep}: a run the runner owns is rewound
 * *on the runner* (this machine has neither its driver nor its worktree).
 *
 * Pair it with {@link remoteReplayStep} rather than reusing it for both.
 * Routing replay through here is exactly what broke remote replay: the
 * runner's retry arm calls `step_retry`, which rejects any step that is not
 * `failed` / `interrupted` / `pending` — so replaying from a completed step
 * failed with "Cannot retry a step in 'completed' status".
 */
export async function remoteRetryStep(input: {
  machineId: string;
  runId: string;
  stepExecutionId: string;
  model: string | null;
  agentKind: string | null;
  effort: EffortLevel | null;
}): Promise<void> {
  await invoke<void>("remote_retry_step", {
    machineId: input.machineId,
    runId: input.runId,
    stepExecutionId: input.stepExecutionId,
    model: input.model,
    agentKind: input.agentKind,
    effort: input.effort,
  });
}

/** The detached twin of {@link replayFromStep} — see {@link remoteRetryStep}. */
export async function remoteReplayStep(input: {
  machineId: string;
  runId: string;
  stepExecutionId: string;
  model: string | null;
  agentKind: string | null;
  effort: EffortLevel | null;
}): Promise<void> {
  await invoke<void>("remote_replay_step", {
    machineId: input.machineId,
    runId: input.runId,
    stepExecutionId: input.stepExecutionId,
    model: input.model,
    agentKind: input.agentKind,
    effort: input.effort,
  });
}

/**
 * Returns `true` when the rejected promise represents a
 * blocking-predecessor error from `step_retry` / `gate_decide`.
 *
 * The backend constructs these as `AppError::validation` with one of
 * the phrases in {@link BLOCKING_ERROR_PHRASES}. Callers should route
 * blocking errors to a `warning` toast instead of an `error` toast,
 * since the user did the right thing (the UI was stale).
 */
export function isBlockingError(err: unknown): boolean {
  const appErr = asAppError(err);
  if (appErr?.kind !== "validation") return false;
  return BLOCKING_ERROR_PHRASES.some((phrase) => appErr.message.includes(phrase));
}

/**
 * Prefix the backend puts on every terminal "the machine is not provisioned"
 * failure (`build_environment_message`). Kept as a constant so the UI's
 * environment hint and the message stay wired to the same string.
 */
export const ENVIRONMENT_ERROR_PREFIX = "Environment not ready";

/**
 * Returns `true` when a step's `error_message` is the backend's terminal
 * environment failure rather than a code failure — i.e. the step did not fail
 * because the change is wrong, it failed because the machine could not run the
 * command at all. The UI uses this to swap "retry the step" framing for "fix
 * the machine" framing, since retrying an unprovisioned box just fails again.
 */
export function isEnvironmentError(errorMessage: string | null | undefined): boolean {
  return (errorMessage ?? "").trimStart().startsWith(ENVIRONMENT_ERROR_PREFIX);
}

/**
 * Pure helper: find the first non-terminal predecessor of `target`
 * in `steps`, ordered by `step_index`. Returns `null` when no
 * predecessor is blocking — used by the UI to decide whether to
 * disable Retry / Approve buttons before the IPC round-trip.
 *
 * The backend enforces the same rule via
 * `assert_no_active_predecessors`; this is the defence-in-depth layer
 * that makes the buttons feel right to the user without a round-trip.
 */
export function findActivePredecessor(
  steps: readonly StepExecution[],
  target: Pick<StepExecution, "id" | "step_index">,
): StepExecution | null {
  for (const s of steps) {
    if (s.id === target.id) continue;
    if (s.step_index >= target.step_index) continue;
    if (
      s.status === "pending" ||
      s.status === "running" ||
      s.status === "verifying" ||
      s.status === "awaiting_gate"
    ) {
      return s;
    }
  }
  return null;
}

/**
 * Same logic as {@link findActivePredecessor} but narrowed to the
 * fields needed to render the gate-modal blocking banner. Pulls the
 * full step list via `step_list_for_run` so the caller doesn't need
 * to thread the steps through the modal props.
 */
export async function listBlockingPredecessor(
  featureId: string,
  target: Pick<StepExecution, "id" | "step_index">,
): Promise<GateBlocker | null> {
  const steps = await listStepsForRun(featureId);
  const found = findActivePredecessor(steps, target);
  if (!found) return null;
  return {
    id: found.id,
    step_id: found.step_id,
    status: found.status,
    step_index: found.step_index,
  };
}